tests/conftest.py

```diff
@@ -1,9 +1,172 @@
 """Fixtures common across tests."""
+import pickle
+from typing import Callable, List
 
+import gym
 import pytest
 import torch
+from stable_baselines3 import PPO
+from stable_baselines3.common.policies import BasePolicy
+from stable_baselines3.common.vec_env import DummyVecEnv, VecNormalize
 
+from imitation.data import rollout
+from imitation.data.types import TrajectoryWithRew
+from imitation.data.wrappers import RolloutInfoWrapper
+from imitation.policies.base import FeedForward32Policy, NormalizeFeaturesExtractor
 from imitation.util import logger
+from imitation.util.networks import RunningNorm
+
+CARTPOLE_ENV_NAME = "CartPole-v1"
+
+
+def load_or_train_ppo(
+    cache_path: str,
+    training_function: Callable[[gym.Env], PPO],
+    venv,
+) -> PPO:
+    try:
+        return PPO.load(cache_path, venv)
+    except OSError:
+        pass
+    except AssertionError:
+        pass
+    expert = training_function(venv)
+    expert.save(cache_path)
+    return expert
+
+
+def load_or_rollout_trajectories(cache_path, policy, venv) -> List[TrajectoryWithRew]:
+    try:
+        with open(cache_path, "rb") as f:
+            return pickle.load(f)
+    except OSError:
+        pass
+    except AssertionError:
+        pass
+    rollout.rollout_and_save(
+        cache_path,
+        policy,
+        venv,
+        rollout.make_sample_until(min_timesteps=2000, min_episodes=57),
+    )
+    with open(cache_path, "rb") as f:
+        return pickle.load(f)  # TODO: not re-loading the trajectory would be nicer here
+
+
+@pytest.fixture(params=[1, 4])
+def cartpole_venv(request) -> gym.Env:
+    num_envs = request.param
+    return DummyVecEnv(
+        [
+            lambda: RolloutInfoWrapper(gym.make(CARTPOLE_ENV_NAME))
+            for _ in range(num_envs)
+        ],
+    )
+
+
+def train_cartpole_expert(cartpole_venv) -> PPO:
+    policy_kwargs = dict(
+        features_extractor_class=NormalizeFeaturesExtractor,
+        features_extractor_kwargs=dict(normalize_class=RunningNorm),
+    )
+    policy = PPO(
+        policy=FeedForward32Policy,
+        policy_kwargs=policy_kwargs,
+        env=VecNormalize(cartpole_venv, norm_obs=False),
+        seed=0,
+        batch_size=64,
+        ent_coef=0.0,
+        learning_rate=0.0003,
+        n_epochs=10,
+        n_steps=64 // cartpole_venv.num_envs,
+    )
+    policy.learn(100000)
```

Comment:
Can we add some check that it meets a minimum level of performance, e.g. average return close to 500?

Our tests assume the expert is actually an expert, so I'd be expect a bunch of spurious failures (and maybe even spurious successes) if expert policy ends up being a dud.

Reply:
Also, does this get run twice because cartpole_venv is parametrized? That seems undesirable if so.

Reply:
The `cartpole_expert_policy` fixture is executed twice. The first time it will train and store a policy. The second time it will just load the policy. So we do not train twice. I tried to ease the confusion be renaming the env parameter in the `train_cartpole_expert` such that there is no name-clash with the `cartpole_venv` fixture.

As for the policy evaluation: should I do some evaluation trajectories or should I check the return during training? What would be the appropriate value for the pendulum environment?

Reply:
Checking return during training is sufficient I think.

[Zoo](https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/logs/benchmark/benchmark.md) reports -172 for `Pendulum-v0` with `ppo`, the best I see is -150 and the worst is around -210. So I'd guesstimate -185 should be enough to catch egregious errors but not be too flaky. Might be worth running a few seeds of PPO on Pendulum to check the mean and variance, then pick something about 2 SD below mean.

Reply:
Turns out that our pendulum expert is far worse than that. I trained it for a couple different seeds and this are the result from  a couple of trial runs. I evaluated using `stable_baselines3.common.evaluation.evaluate_policy`:

Mean Reward | Std over evaluation episodes
-------------------|---------------------
-232.0 | 315.3
-320.5 | 450.7
-621.3 | 314.6
-902.8 | 342.3
-421.4 | 339.2
-423.5 | 340.6
-528.8 | 334.0
-906.5 | 417.0
-500.9 | 382.0
-611.7 | 372.9

Mean over means: -547, stdev over means: 212

Turns out we do not use exactly the hyperparameters from Zoo. I am currently running a test with the parameters from there.

Reply:
Results with the Zoo hyperparameters:

Mean Reward | Std over evaluation episodes
-------------------|---------------------
-368.2 | 389.6
-221.2 | 98.1
-132.5 | 102.7
-161.3 | 82.7
-183.3 | 142.6
-198.8 | 111.0
-152.9 | 71.5
-122.0 | 71.6
-342.6 | 142.1
-185.6 | 64.5

Mean over means: -207, stdev over means: 80

This does not quite match the results from the Zoo. However, note that the zoo uses v0 and we use v1 of the environment.

I will go with the median of -184.5 your guesstimate of -185 was not too bad! I will use that threshold now.

Reply:
Thanks for running these tests, and spotting the hyperparam difference.

If we fail if return is below median, this will fail 50% of the time? Do we want to automatically re-run it a certain number of times? Alternatively it's possible that boosting the number of `total_timesteps` moderately would improve consistency.